// ngx-trim/fesm2015/ngx-trim.js
import { Directive, ElementRef, forwardRef, HostListener, Renderer2, Input, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR, FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ TRIM_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    // tslint:disable-next-line:no-forward-ref
    useExisting: forwardRef(() => NgxTrimDirective),
    multi: true
};
class NgxTrimDirective {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     */
    constructor(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
    }
    /**
     * @param {?} _
     * @return {?}
     */
    _onChange(_) { }
    /**
     * @return {?}
     */
    _onTouched() { }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (value !== undefined && value !== null) {
            this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.onEvent = this.onEvent || 'focusout';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _onKeyUp(event) {
        if (this.onEvent === event.type) {
            const /** @type {?} */ element = /** @type {?} */ (event.target);
            const /** @type {?} */ val = element.value.trim();
            this.writeValue(val);
            this._onChange(val);
        }
    }
}
NgxTrimDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'input[ngxTrim]',
                providers: [TRIM_VALUE_ACCESSOR]
            },] },
];
/** @nocollapse */
NgxTrimDirective.ctorParameters = () => [
    { type: Renderer2, },
    { type: ElementRef, },
];
NgxTrimDirective.propDecorators = {
    "onEvent": [{ type: Input, args: ['ngxTrim',] },],
    "_onKeyUp": [{ type: HostListener, args: ['keyup', ['$event'],] }, { type: HostListener, args: ['focusout', ['$event'],] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxTrimModule {
}
NgxTrimModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, FormsModule],
                declarations: [NgxTrimDirective],
                exports: [NgxTrimDirective]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxTrimDirective, NgxTrimModule };
