// ngx-trim/bundles/ngx-trim.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-trim', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (factory((global['ngx-trim'] = {}),global.ng.core,global.ng.forms,global.ng.common));
}(this, (function (exports,core,forms,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ TRIM_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        // tslint:disable-next-line:no-forward-ref
        useExisting: core.forwardRef(function () { return NgxTrimDirective; }),
        multi: true
    };
    var NgxTrimDirective = (function () {
        function NgxTrimDirective(_renderer, _elementRef) {
            this._renderer = _renderer;
            this._elementRef = _elementRef;
        }
        /**
         * @param {?} _
         * @return {?}
         */
        NgxTrimDirective.prototype._onChange = /**
         * @param {?} _
         * @return {?}
         */
            function (_) { };
        /**
         * @return {?}
         */
        NgxTrimDirective.prototype._onTouched = /**
         * @return {?}
         */
            function () { };
        /**
         * @param {?} fn
         * @return {?}
         */
        NgxTrimDirective.prototype.registerOnChange = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this._onChange = fn;
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        NgxTrimDirective.prototype.registerOnTouched = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this._onTouched = fn;
            };
        /**
         * @param {?} value
         * @return {?}
         */
        NgxTrimDirective.prototype.writeValue = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (value !== undefined && value !== null) {
                    this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
                }
            };
        /**
         * @return {?}
         */
        NgxTrimDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.onEvent = this.onEvent || 'focusout';
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NgxTrimDirective.prototype._onKeyUp = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.onEvent === event.type) {
                    var /** @type {?} */ element = (event.target);
                    var /** @type {?} */ val = element.value.trim();
                    this.writeValue(val);
                    this._onChange(val);
                }
            };
        NgxTrimDirective.decorators = [
            { type: core.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: 'input[ngxTrim]',
                        providers: [TRIM_VALUE_ACCESSOR]
                    },] },
        ];
        /** @nocollapse */
        NgxTrimDirective.ctorParameters = function () {
            return [
                { type: core.Renderer2, },
                { type: core.ElementRef, },
            ];
        };
        NgxTrimDirective.propDecorators = {
            "onEvent": [{ type: core.Input, args: ['ngxTrim',] },],
            "_onKeyUp": [{ type: core.HostListener, args: ['keyup', ['$event'],] }, { type: core.HostListener, args: ['focusout', ['$event'],] },],
        };
        return NgxTrimDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxTrimModule = (function () {
        function NgxTrimModule() {
        }
        NgxTrimModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule, forms.FormsModule],
                        declarations: [NgxTrimDirective],
                        exports: [NgxTrimDirective]
                    },] },
        ];
        return NgxTrimModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.NgxTrimDirective = NgxTrimDirective;
    exports.NgxTrimModule = NgxTrimModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
