// ngx-trim/esm5/lib/ngx-trim.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, forwardRef, HostListener, Renderer2 } from '@angular/core';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var /** @type {?} */ TRIM_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    // tslint:disable-next-line:no-forward-ref
    useExisting: forwardRef(function () { return NgxTrimDirective; }),
    multi: true
};
var NgxTrimDirective = /** @class */ (function () {
    function NgxTrimDirective(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
    }
    /**
     * @param {?} _
     * @return {?}
     */
    NgxTrimDirective.prototype._onChange = /**
     * @param {?} _
     * @return {?}
     */
    function (_) { };
    /**
     * @return {?}
     */
    NgxTrimDirective.prototype._onTouched = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxTrimDirective.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxTrimDirective.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onTouched = fn;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxTrimDirective.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value !== undefined && value !== null) {
            this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
        }
    };
    /**
     * @return {?}
     */
    NgxTrimDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.onEvent = this.onEvent || 'focusout';
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NgxTrimDirective.prototype._onKeyUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.onEvent === event.type) {
            var /** @type {?} */ element = /** @type {?} */ (event.target);
            var /** @type {?} */ val = element.value.trim();
            this.writeValue(val);
            this._onChange(val);
        }
    };
    NgxTrimDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'input[ngxTrim]',
                    providers: [TRIM_VALUE_ACCESSOR]
                },] },
    ];
    /** @nocollapse */
    NgxTrimDirective.ctorParameters = function () { return [
        { type: Renderer2, },
        { type: ElementRef, },
    ]; };
    NgxTrimDirective.propDecorators = {
        "onEvent": [{ type: Input, args: ['ngxTrim',] },],
        "_onKeyUp": [{ type: HostListener, args: ['keyup', ['$event'],] }, { type: HostListener, args: ['focusout', ['$event'],] },],
    };
    return NgxTrimDirective;
}());
export { NgxTrimDirective };
function NgxTrimDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxTrimDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxTrimDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgxTrimDirective.propDecorators;
    /** @type {?} */
    NgxTrimDirective.prototype.onEvent;
    /** @type {?} */
    NgxTrimDirective.prototype._renderer;
    /** @type {?} */
    NgxTrimDirective.prototype._elementRef;
}
