// ngx-trim/esm2015/lib/ngx-trim.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, forwardRef, HostListener, Renderer2 } from '@angular/core';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
const /** @type {?} */ TRIM_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    // tslint:disable-next-line:no-forward-ref
    useExisting: forwardRef(() => NgxTrimDirective),
    multi: true
};
export class NgxTrimDirective {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     */
    constructor(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
    }
    /**
     * @param {?} _
     * @return {?}
     */
    _onChange(_) { }
    /**
     * @return {?}
     */
    _onTouched() { }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (value !== undefined && value !== null) {
            this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.onEvent = this.onEvent || 'focusout';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _onKeyUp(event) {
        if (this.onEvent === event.type) {
            const /** @type {?} */ element = /** @type {?} */ (event.target);
            const /** @type {?} */ val = element.value.trim();
            this.writeValue(val);
            this._onChange(val);
        }
    }
}
NgxTrimDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'input[ngxTrim]',
                providers: [TRIM_VALUE_ACCESSOR]
            },] },
];
/** @nocollapse */
NgxTrimDirective.ctorParameters = () => [
    { type: Renderer2, },
    { type: ElementRef, },
];
NgxTrimDirective.propDecorators = {
    "onEvent": [{ type: Input, args: ['ngxTrim',] },],
    "_onKeyUp": [{ type: HostListener, args: ['keyup', ['$event'],] }, { type: HostListener, args: ['focusout', ['$event'],] },],
};
function NgxTrimDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxTrimDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxTrimDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgxTrimDirective.propDecorators;
    /** @type {?} */
    NgxTrimDirective.prototype.onEvent;
    /** @type {?} */
    NgxTrimDirective.prototype._renderer;
    /** @type {?} */
    NgxTrimDirective.prototype._elementRef;
}
